pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -1116,18 +1170,22 @@ func addMetadataByPrecedence(metadata map[string]string, addedMetadata map[strin
 // If there is no timeout for the finally TaskRun, returns 0.
 // If pipeline level timeouts have already been exceeded, returns 1 second.
 func getFinallyTaskRunTimeout(ctx context.Context, pr *v1beta1.PipelineRun, rpt *resources.ResolvedPipelineTask, c clock.PassiveClock) *metav1.Duration {
```

Comment:
I'd like to propose the following strategy:

- Set a taskrun/run's timeout to the timeout of the pipeline task
- when a pipelinerun has hit its tasks timeout, cancel all the non-finally tasks
- when a pipelinerun has hit its finally timeout, cancel all the finally tasks
- when a pipelinerun has hit its pipeline timeout, cancel all the tasks

the functions `getFinallyTaskRunTimeout` and `getTaskRunTimeout` can go away entirely

Reply:
That's basically what we're doing in this PR, isn't it? With the caveat that in `calculateTaskRunTimeout`, we're also checking if the current time is past the appropriate `.timeouts` value (calculating based on `StartTime` or `FinallyStartTime` for `.timeouts.tasks` and `.timeouts.finally` respectively, falling back to `.timeouts.pipeline` when `.timeouts.tasks` or `.timeouts.finally` is unset) to do the admittedly hacky 1s timeout for "too late to launch" tasks, then falling back on `PipelineTask.Timeout`, so it's slightly more complicated.

So I guess you'd like to get rid of the 1s timeout for "too late to launch" tasks, since that's the only real difference from what you're proposing. I do not care for the 1s thing, so if you have any suggestions for an alternative that retains the general behavior we've got currently, I'm more than open to it. =) I'll also poke at the code to see if anything comes to mind myself.

Reply:
Ok, so I think we could switch from the "make this fail right away" 1s timeout to returning -1 as the timeout duration, and instead marking the task as skipped (with a new skipped reason for being past timeout) if we're past the latest possible launch time for it. That seems like a good idea to me, but it still would require `calculateTaskRunTimeout`, `getFinallyTaskRunTimeout`, and `getTaskRunTimeout`, _and_ would be a decent refactor. So I don't think that would simplify the code in the context of this PR, and it feels like it's a separate issue/change.

Reply:
I think this PR is sort of doing what I described in that it's cancelling the taskruns when the appropriate timeout is reached. I'm suggesting removing the logic that takes `pipelinerun.spec.timeouts` into account when setting `taskrun.timeout`. I don't think it should be needed anymore, and I would think it will continue to lead to the type of race conditions this PR is attempting to fix.

Reply:
The reason that logic is there is to deal with the cases where we go to create a `TaskRun`/`Run` but we're already past the `.timeouts.tasks` or `.timeouts.finally` timeout for that `TaskRun`/`Run` - that's what I was talking about redoing separately to not create a `TaskRun`/`Run` with a `1s` timeout in those scenarios, but instead to mark that task as skipped. Which, like I said, I think should be in a separate PR.

Reply:
ok, I think I understand. I'm wondering if instead we can change the logic so the PipelineRun doesn't continue on to `runNextSchedulableTask` if it's reached its `pipeline` timeout or `finally` timeout? I guess we'd have to write some function like `runFinallyTasks` for if it reaches its `tasks` timeout.
basically, within `reconcile`, it feels like we need a way to transition into the `finally` stage or something, similar to [finishReconcileUpdateEmitEvents](https://github.com/tektoncd/pipeline/blob/deb0fec6e861ed9c74c7f43791b6e52823a62e0d/pkg/reconciler/pipelinerun/pipelinerun.go#L263)

Reply:
Yeah, I'd have to think about it - this is why I think it's a separate issue. =)

Reply:
Hm I am not sure we can push this off for a follow-up PR, because I think the strategy being used here introduces a new race condition (for timeouts.tasks and timeouts.finally) even as it's fixing an existing one

Reply:
What's the race condition?

Reply:
Ah, yeah, that is possible, though I think significantly less likely than the current race condition. I'd still prefer to immediately follow up with revamping the "don't run after timeout" logic in a separate PR, just to keep this one from getting more complex, but I'll defer to your judgment there.

Reply:
I think it'll happen whenever a user configures `timeouts.tasks`. Could you maybe experiment with another commit on top of this one? While I want to avoid making this PR more complex, refactoring the behavior we're talking about could be good to do first, as it'll allow us to remove the functions `getFinallyTaskRunTimeout` and `getTaskRunTimeout` entirely.

Reply:
I'm playing around with having `ResolvedPipelineTask.Skip(facts)` mark a task as skipped if it's past the `tasks` or `finally` timeout right now, which feels like the right way to do it to me. =)

Reply:
@lbernick Take a look at https://github.com/tektoncd/pipeline/pull/5134/commits/daf32f408d70c417da898effbdba465da7c89a14 - that's my first draft of the skipping approach.